File: http/src/main/java/org/http4s/blaze/http20/FrameHandler.java
//package org.http4s.blaze.http20;
//
//import java.nio.ByteBuffer;
//
//public interface FrameHandler {
//
//    boolean onDataFrame(ByteBuffer data, int streamId, boolean isLast);
//
//    boolean onHeadersFrame(ByteBuffer data,
//                           int streamId,
//                           int streamDep,
//                           boolean exclusive,
//                       int priority,
//                       boolean endHeaders,
//                       boolean endStream);
//
//    Boolean onPriorityFrame(int streamId, int streamDep, boolean exclusive);
//
//    Boolean onRstStreamFrame(int streamId, int code);
//
//    // Each setting is handled one at a time. false signals the decoding process to halt
//    Boolean handleSetting(int id, long value);
//
//    Boolean onPushPromiseFrame(int streamId, int promisedId, boolean endHeaders, ByteBuffer data);
//
//    Boolean onPingFrame(byte []data, boolean ack);
//
//    Boolean onGoAwayFrame(int lastStream, long errorCode, ByteBuffer debugData);
//
//    Boolean onWindowUpdateFrame(int streamId, int sizeIncrement);
//
//    Boolean onContinuationFrame(ByteBuffer data, int streamId, boolean endHeaders);
//
//    // For handling unknown stream frames
//    Boolean onUnknownFrame(int tpe, int streamId, byte flags, ByteBuffer data);
//
//}

File: http/src/main/scala/org/http4s/blaze/http/http20/FrameHandler.scala
package org.http4s.blaze.http.http20

import java.nio.ByteBuffer

trait FrameHandler {

  def inHeaderSequence(): Boolean

  def onDataFrame(streamId: Int, isLast: Boolean, data: ByteBuffer, flowSize: Int): DecoderResult

  def onHeadersFrame(streamId: Int,
                    streamDep: Int,
                    exclusive: Boolean,
                     priority: Int,
                  end_headers: Boolean,
                   end_stream: Boolean,
                   buffer: ByteBuffer): DecoderResult

  def onPriorityFrame(streamId: Int, streamDep: Int, exclusive: Boolean, priority: Int): DecoderResult

  def onRstStreamFrame(streamId: Int, code: Int): DecoderResult

  def onSettingsFrame(ack: Boolean, settings: Seq[Setting]): DecoderResult

  def onPushPromiseFrame(streamId: Int, promisedId: Int, end_headers: Boolean, data: ByteBuffer): DecoderResult

  def onPingFrame(data: Array[Byte], ack: Boolean): DecoderResult

  def onGoAwayFrame(lastStream: Int, errorCode: Long, debugData: ByteBuffer): DecoderResult

  def onWindowUpdateFrame(streamId: Int, sizeIncrement: Int): DecoderResult

  def onContinuationFrame(streamId: Int, endHeaders: Boolean, data: ByteBuffer): DecoderResult

  // For handling unknown stream frames
  def onExtensionFrame(tpe: Int, streamId: Int, flags: Byte, data: ByteBuffer): DecoderResult

  ///////////// Error Handling //////////////////////////////////
}

